File: smmr_24_25_class_2.R
####################################################################
###### SMMR by Ilya S. Slabolitskiy and Ekaterina A. Simonova ######
############################# Class 2 ##############################
####################################################################



# install.packages('mfx')
library('mfx') # margibal effects
# disable scientific notation
options(scipen = 999)


# problem 1 from class 2
# data simulation
n_obs <- 1000
beta_0 <- 2
beta_1 <- -3
epsilon <- rlogis(n = n_obs)
x <- rbinom(n = n_obs, size = 1, prob = 0.75)
y <- as.numeric(beta_0 + beta_1 * x + epsilon > 0)
df <- data.frame('y' = y, 'x' = x)

# logit model estimation
logit_1 <- glm(formula = y ~ x, data = df, family = binomial(link = 'logit'))
logit_1$coefficients

# numerical answer vs analytical answer
n_00 <- sum((y == 0) & (x == 0))
n_01 <- sum((y == 0) & (x == 1))
n_10 <- sum((y == 1) & (x == 0))
n_11 <- sum((y == 1) & (x == 1))
log(n_10 / n_00)
log(n_11 * n_00 / (n_01 * n_10))

# clear environment
rm(list = ls())


# marginar effects and interpretation
# data simulation
n_obs <- 1000
beta_0 <- 2
beta_1 <- -3
beta_2 <- 3
epsilon <- rlogis(n = n_obs)
x_1 <- rexp(n = n_obs, rate = 1)
x_2 <- rbinom(n = n_obs, size = 1, prob = 0.75)
y <- as.numeric(beta_0 + beta_1 * x_1 + beta_2 * x_2 + epsilon > 0)
df <- data.frame('y' = y, 'x_1' = x_1, 'x_2' = x_2)

# logit model estimation
logit_2 <- glm(y ~ x_1 + x_2, data = df, family = binomial(link = 'logit'))
logit_2$coefficients
# marginal effects
logitmfx(formula = y ~ x_1 + x_2, data = df, atmean = FALSE)


# custom log-likelihood function and numerical optimization
LogL <- function(beta)
{
  beta0 <- beta[1]
  beta1 <- beta[2]
  beta2 <- beta[3]
  
  cond0 <- y == 0
  cond1 <- y == 1
  
  y_star <- beta0 + beta1 * x_1 + beta2 * x_2
  log_lik <- sum(log(1 - plogis(q = y_star))[cond0]) +
             sum(log(plogis(q = y_star))[cond1])
  
  return(log_lik)
}

logL_optim <- optim(par = rep(0, 3),
                    fn = LogL,
                    method = 'Nelder-Mead',
                    hessian = TRUE,
                    lower = rep(-Inf, 3),
                    upper = rep(Inf, 3),
                    control = list(maxit = 10000,
                                   abstol = 1e-10,
                                   fnscale = -1))
logL_optim$par
logL_optim$value
logL_optim$hessian
-solve(logL_optim$hessian)

File: smmr_24_25_class_2.py
import numpy as np
import pandas as pd
import statsmodels.api as sm
from scipy import stats
from scipy.optimize import minimize
from scipy.special import expit, logit

# Disable scientific notation
pd.set_option('display.float_format', lambda x: '%.5f' % x)

# Problem 1: Simple Logistic Regression
# =====================================

# Data simulation
np.random.seed(42)
n_obs = 1000
beta_0 = 2
beta_1 = -3

epsilon = stats.logistic.rvs(size=n_obs)
x = stats.bernoulli.rvs(0.75, size=n_obs)
y = (beta_0 + beta_1*x + epsilon > 0).astype(int)
df = pd.DataFrame({'y': y, 'x': x})

# Logit model estimation
X = sm.add_constant(df[['x']])
logit_model = sm.Logit(df['y'], X).fit()
print("Problem 1 Coefficients:")
print(logit_model.params)

# Analytical solution
n00 = ((y == 0) & (x == 0)).sum()
n01 = ((y == 0) & (x == 1)).sum()
n10 = ((y == 1) & (x == 0)).sum()
n11 = ((y == 1) & (x == 1)).sum()

analytical_beta0 = np.log(n10/n00)
analytical_beta1 = np.log((n11*n00)/(n01*n10))

print("\nAnalytical Solutions:")
print(f"Beta0: {analytical_beta0:.4f}")
print(f"Beta1: {analytical_beta1:.4f}")

# Problem 2: Logistic Regression with Marginal Effects
# ====================================================

# Data simulation
np.random.seed(42)
n_obs = 1000
beta_0 = 2
beta_1 = -3
beta_2 = 3

epsilon = stats.logistic.rvs(size=n_obs)
x1 = stats.expon.rvs(size=n_obs)
x2 = stats.bernoulli.rvs(0.75, size=n_obs)
y = (beta_0 + beta_1*x1 + beta_2*x2 + epsilon > 0).astype(int)
df = pd.DataFrame({'y': y, 'x1': x1, 'x2': x2})

# Logit model estimation
X = sm.add_constant(df[['x1', 'x2']])
logit_model2 = sm.Logit(df['y'], X).fit()
print("\nProblem 2 Coefficients:")
print(logit_model2.params)

# Marginal effects (average marginal effects)
predicted_probs = logit_model2.predict(X)
marginal_effects = pd.DataFrame({
    'x1': logit_model2.params['x1'] * predicted_probs * (1 - predicted_probs),
    'x2': logit_model2.params['x2'] * predicted_probs * (1 - predicted_probs)
})

print("\nMarginal Effects (Average):")
print(marginal_effects.mean())

# Custom Log-Likelihood Function
# =============================

def log_likelihood(beta, y, x1, x2):
    Xb = beta[0] + beta[1]*x1 + beta[2]*x2
    log_lik = np.sum(y * np.log(expit(Xb))) + np.sum((1-y) * np.log(1 - expit(Xb)))
    return -log_lik  # Negative for minimization

# Numerical optimization
result = minimize(log_likelihood,
                  x0=np.zeros(3),
                  args=(df['y'], df['x1'], df['x2']),
                  method='Nelder-Mead')

print("\nOptimization Results:")
print("Coefficients:", result.x)
print("Log-Likelihood:", -result.fun)

# Calculate Hessian-based standard errors
hessian = result.hess_inv(np.eye(3)) if result.hess_inv is not None else np.zeros((3,3))
std_errors = np.sqrt(np.diag(hessian))
print("\nStandard Errors:", std_errors)